sdk/protocol-plugins/src/plugins/common/helpers/aaveV3Like/AAVEv3LikeBaseProtocolPlugin.ts

```diff
@@ -0,0 +1,243 @@
+import { CommonTokenSymbols, IToken } from '@summerfi/sdk-common/common'
+import { ICollateralInfo } from '@summerfi/sdk-common/protocols'
+
+type AssetsList = Awaited<ReturnType<AAVEv3LikeBaseProtocolPlugin['_getAssetsList']>>
+type Asset = AssetsList extends (infer U)[] ? U : never
+
+/**
+ * @class AAVEv3BaseProtocolPlugin
+ * @description Base class for AAVEv3 protocol plugins, it contains common functionality to
+ * fetch data from forks of AAVEv3 protocol.
+ */
+export abstract class AAVEv3LikeBaseProtocolPlugin extends BaseProtocolPlugin {
+  abstract readonly protocolName: AllowedProtocolNames
+  private _assetsList: Maybe<AssetsList> = undefined
+
+  /** PROTECTED */
+
+  /**
+   * Fetches the contract definition for the given contract name.
+   * @param contractName The name of the contract to fetch the definition for.
+   * @returns The contract definition for the given contract name.
+   *
+   * To be implemented by each specific protocol plugin.
+   */
+  protected abstract _getContractDef(contractName: string): AAVEv3LikeAbiInfo
+
+  /**
+   * Initializes the assets list if it hasn't been initialized yet.
+   *
+   * To be called before fetching data from the assets list, typically from `getLendingPool` and similar
+   * methods.
+   */
+  protected async _inititalizeAssetsListIfNeeded() {
+    if (this._assetsList) {
+      return
+    }
+
+    this._assetsList = await this._getAssetsList()
+  }
+
+  /**
+   * Fetches the assets list for the protocol.
+   * @returns The assets list for the protocol.
+   *
+   * This function must exist as the return type of the data builder is inferred from the return
+   * type of this function.
+   */
+  protected async _getAssetsList() {
+    try {
+      const _ctx = {
+        ...this.ctx,
+        getContractDef: this._getContractDef,
+      }
+      const builder = await new AaveV3LikeProtocolDataBuilder(_ctx, this.protocolName).init()
+      return await builder
+        .addPrices()
+        .addReservesCaps()
+        .addReservesConfigData()
+        .addReservesData()
+        .addEmodeCategories()
+        .build()
+    } catch (e) {
+      throw new Error(`Could not fetch/build assets list for AAVEv3: ${JSON.stringify(e)}`)
+    }
+  }
+
+  /**
+   * Fetches the asset from the assets list for the given token and emode.
+   * @param token  The token to fetch the asset for.
+   * @param emode  The emode to fetch the asset for.
+   * @returns  The asset for the given token and emode.
+   */
+  protected async _getAssetFromToken(token: IToken, emode: bigint): Promise<Asset> {
+    if (!this._assetsList) {
+      throw new Error('Assets list not initialized')
+    }
+
+    const assetsList = filterAssetsListByEMode(this._assetsList, emode)
+
+    const asset = assetsList.find((asset: Asset) => token.equals(asset.token))
+    if (!asset) {
+      throw new Error(`Asset not found for token ${token}`)
+    }
+
+    return asset
+  }
```

Comment:
Another function which I don't really understand what it does? Why it filers by emode, what gets filtere?

Reply:
This is partially coming from the previous implementation. We need to get the right collateral/debt info for the given assets and given eMode. We fetch ALL assets from the protocol, then filter by asset type (symbol) then by eMode and we get the final asset that is used to fill in the LendingPoolInfo structure